DRA kubeletplugin: turn helper into wrapper

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The goal is to simplify writing DRA drivers. This is also a first step towards supporting seamless upgrades.

DRA drivers no longer need to implement the kubelet plugin API directly. Instead, the helper wraps an implementation of an interface. The helper then provides common functionality:

- retrieve and validate ResourceClaims
- serialize gRPC calls (enabled by default, can be opted out)
- gRPC logging

The definition of that interface is meant to be comprehensive enough that a correct DRA driver can be implemented by following the documentation of the package, without having to cross-reference KEPs.

#### Which issue(s) this PR fixes:
Related-to: https://github.com/kubernetes/kubernetes/issues/128964

#### Special notes for your reviewer:

The DRAPlugin interface used to be the abstract API of the helper. Now it's what the DRA driver kubelet plugin needs to implement. The helper is a concrete Server struct with no exported fields. It only exports the methods that drivers need when using the helper.

While at it, support for the v1alpha4 API gets removed from the helper, which implies removing the corresponding E2E tests. The kubelet implementation will be dropped separately.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @byako @elezar @bart0sh

pohly committed:
DRA kubeletplugin: turn helper into wrapper

The goal is to simplify writing DRA drivers. This is also a first step towards
supporting seamless upgrades.

DRA drivers no longer need to implement the kubelet plugin API
directly. Instead, the helper wraps an implementation of an interface. The
helper then provides common functionality:

- retrieve and validate ResourceClaims
- serialize gRPC calls (enabled by default, can be opted out)
- gRPC logging

The definition of that interface is meant to be comprehensive enough that a
correct DRA driver can be implemented by following the documentation of the
package, without having to cross-reference KEPs.

The DRAPlugin interface used to be the abstract API of the helper. Now it's
what the DRA driver kubelet plugin needs to implement. The helper is a concrete
Server struct with no exported fields. It only exports the methods that
drivers need when using the helper.

While at it, support for the v1alpha4 API gets removed from the helper, which
implies removing the corresponding E2E tests. The kubelet implementation will
be dropped separately.

pohly committed:
DRA kubeletplugin: revise socket path handling

When supporting rolling updates, we cannot use the same fixed socket paths for
old and new pod. With the revised API, the caller no longer specifies the full
socket paths, only directories. The logic about how to name sockets then can be
in the helper.

While at it, avoid passing a context to the gRPC helper code when
all that the helper code needs is a logger. That leads to confusion
about whether cancellation has an effect.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -21,56 +21,128 @@ import (
+	// subrequest name if there is one.
+	//
+	// If empty, the device is associated with all requests.
+	Requests []string
```
@mortent: this is what the helper expects from the DRA driver. Note that it doesn't ask DRA driver authors to strip the subrequest name. Instead, the helper will do that before returning this as result to the kubelet.

Looks okay to you?

mortent replied:
Yeah, that makes sense to me. Since we don't allow references to subrequests in the pod/container specs, the logic that determines which requests should be available to pods/containers must strip away the subrequest (if specified). Doing this in the helper seems like the right approach.

pohly committed:
DRA e2e: more flexible socket listening

Instead of hard-coding two instances of the hostpathplugin which listen on
certain socket paths, the hostpathplugin now gets started through Pod exec as
needed. The advantage is that the helper code is in charge of socket naming,
just like it would be in real deployment.

One nuisance is that exec.StreamWithContext always complains in copyFromStdout
and copyFromStderr when the remote hostpathplugin gets killed via context
cancellation:

    E0312 11:56:31.637669  289446 v2.go:167] "Unhandled Error" err="next reader: read tcp [::1]:59006->[::1]:6444: use of closed network connection" logger="UnhandledError"
    E0312 11:56:31.637749  289446 v2.go:150] "Unhandled Error" err="next reader: read tcp [::1]:59006->[::1]:6444: use of closed network connection" logger="UnhandledError"

These can be ignored.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/namespacedobject.go:
```diff
@@ -0,0 +1,50 @@
+type NamespacedObject struct {
+	types.NamespacedName
+	UID types.UID
+}
```
This is the second PR where I need such a type. It's a verbatim copy.

Long term I want to consolidate and move it to `k8s.io/apimachinery/pkg/types`, but not now,

pohly commented:
/assign @byako @nojnhuh @elezar 

Let's merge this once there is a sufficient number of +1 from consumers of this package - ideally within the next few days.

/hold

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -21,56 +21,128 @@ import (
+	// CDIDeviceIDs lists all CDI devices associated with this DRA device.
+	// Each ID must be of the form "<vendor ID>/<class>=<unique name>".
+	// May be empty.
+	CDIDeviceIDs []string
```
Question: Does it make sense to wrapp the id's in a type for parsing etc?

pohly replied:
Nothing in Kubernetes parses these IDs. If https://github.com/cncf-tags/container-device-interface had a type for it *and* we could import that package, we could use that type. But that is a lot of ifs for a little bit of extra type safety, so perhaps let's stick with `string`?

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -94,63 +166,71 @@ func GRPCVerbosity(level int) Option {
+//
+// The default is <driver name>-reg.sock. When rolling updates are enabled (not supported yet),
+// it is <driver name>-<uid>-reg.sock.
+func RegistrarSocketName(name string) Option {
```
Is `RegistrarSocketFilename` more accurate?

Also, who is responsible for determining the `uuid` mentioned in the docs?

elezar replied:
OK. Thanks. That's clearer. Note that the docstring refers to `RegistrarSocketName` which we have renamed, correct?

pohly replied:
I also renamed it in https://github.com/kubernetes/kubernetes/pull/129832.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -173,17 +253,8 @@ func GRPCStreamInterceptor(interceptor grpc.StreamServerInterceptor) Option {
-
 // NodeV1beta1 explicitly chooses whether the DRA gRPC API v1beta1
-// gets enabled.
+// gets enabled. True by default.
```
Is a boolean flexible enough for this? How does this change as we add new API versions? What happens when this is set to `false`?

pohly replied:
When explicitly set to false, the server cannot start at the moment because there is no other API version left.

I could have removed it, but then I would have to bring it back if we ever add another alternative.

pohly replied:
I added the same comment as for the listener options:
> This is used in Kubernetes for end-to-end testing. The default should
> be fine for DRA drivers.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -223,37 +294,51 @@ func NodeUID(nodeUID types.UID) Option {
-// draPlugin combines the kubelet registration service and the DRA node plugin
-// service.
-type draPlugin struct {
+// Helper combines the kubelet registration service and the DRA node plugin
```
Is there a less generic name than `Helper`?

pohly replied:
Suggestions? :sweat_smile: 

Remember that the full name is `kubeletplugin.Helper`. That is a bit more descriptive.

We could rename to "Adapter" or "Bridge", but those are also generic.

elezar replied:
Is `kubeletplugin.RegistrationHelper` more accurate? (I assuming this does more than just registration though).

What about `kubeletplugin.DRAHelper` since DRA is not readable at the callsite at all?

pohly replied:
`RegistrationHelper` is less accurate because it also implements the DRA plugin gRPC API.

`DRAHelper` may help, but is at the wrong place. A better solution is:
```
import drakubeletplugin "k8s.io/dynamic-resource-allocation/kubeletplugin"
...
...  drakubeletplugin.Helper ...
```

elezar replied:
Yes, that does look better.

l'm not going to block on this name though.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -94,63 +166,71 @@ func GRPCVerbosity(level int) Option {
+// RegistrarListener configures how to create the registrar socket.
+// The default is to remove the file if it exists and to then
+// create a socket.
+func RegistrarListener(listen func(ctx context.Context, path string) (net.Listener, error)) Option {
```
The default is the code in `startGRPCServer`. There is no separate standalone function.

Normal drivers should not need this. It's for testing in test/e2e/deploy.go.

pohly replied:
Added comment:
> This is used in Kubernetes for end-to-end testing. The default should
> be fine for DRA drivers.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -223,37 +294,51 @@ func NodeUID(nodeUID types.UID) Option {
 	}
 }
 
+// Serialize overrides whether the helper serializes the prepare and unprepare
```
I'm not aware of conventions around naming something like this. I prefer `Serialize` because `Serialize(false)` is unambiguous: "don't serialize!". With `MustSerialize(false)` the helper might still do it even though it's not required.

nojnhuh commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -21,56 +21,128 @@ import (
+	// the corresponding ResourceClaim. For each claim, preparation
+	// can be either successful (no error set in the per-ResourceClaim PrepareResult)
+	// or can be reported as failed.
+	PrepareResourceClaims(ctx context.Context, claims []*resourceapi.ResourceClaim) (result map[types.UID]PrepareResult, err error)
```
I [updated the dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver/commit/74c75cffb404147375c60e4c5e9d90adf945b35d) to use the new helper and everything was pretty straightforward.

The only slightly annoying part was dealing with this new return type. The example driver stores checkpoints on disk with serialized [Devices](https://pkg.go.dev/k8s.io/kubelet@v0.33.0-beta.0/pkg/apis/dra/v1beta1#Device) defined by the kubelet protobuf API that it used to return from `NodePrepareResources`. Since the helper wants the `Device` type defined in this package, I added a [simple loop to convert](https://github.com/nojnhuh/dra-example-driver/blob/f803b6e0949b6387f9fade83a1c46cc9cad581e3/cmd/dra-example-kubeletplugin/driver.go#L120-L128) the protobuf type to the helper type to avoid having to change any assumptions about what might be stored on disk and since that seems to match how the helper's `NodePrepareResources` below works. For other drivers in a similar position, would this kind of loop be a recommended way to handle that?

pohly replied:
That the example driver uses the protobuf encoding as its own on-disk format is a bit problematic. It doesn't "own" that format. What if we extend it upstream? That aside, how would it deal with transitions in the gRPC API? Suppose it stores in v1beta1 and we transition to v1beta2 with new fields. How does it store those? How does it convert? Does it support downgrades in that case? 

I think using the same encoding on disk and in the interaction with the kubelet was a shortcut. It would be cleaner to accept that the two are different and plan accordingly. Your proposed loop basically does that, so it sounds like the right approach even if it might be the final solution.

bg-furiosa commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -21,56 +21,128 @@ import (
+	//
+	// The conventions for returning one overall error and several per-ResourceClaim
+	// errors are the same as in PrepareResourceClaims.
+	UnprepareResourceClaims(ctx context.Context, claims []NamespacedObject) (result map[types.UID]string, err error)
```
the return type `result map[types.UID]string, err error `, it seems unclear.
No idea what string should be.
How about defining and using a type like ErrMsg instead of string.

pohly replied:
What if we simply use `error` instead? That allows "normal" error handling in the DRA driver. Serializing as string then happens only when needed (= when returning results via gRPC). I switched to that. It made the test driver simpler and matches how gRPC handles error return values.

I also remembered that you asked on Slack about dynamically creating the CDI spec files and added something about it to PrepareResourceClaims:

```
	// It is possible to create the CDI spec files which define the CDI devices
	// on-the-fly in PrepareResourceClaims. UnprepareResourceClaims then can
	// remove them. Container runtimes may cache CDI specs but must reload
	// files in case of a cache miss. To avoid false cache hits, the unique
	// name in the CDI device ID should not be reused. A DRA driver can use
	// the claim UID for it.
```

bg-furiosa replied:
much looks better!

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/doc.go:
```diff
@@ -16,4 +16,39 @@ limitations under the License.
+// Therefore the DaemonSet must not set `maxSurge` to a value larger than
+// zero. With the default `maxSurge: 0`, updating the DaemonSet of the driver
+// will first shut down the old driver Pod, then start the replacement.
+// This leads to a short downtime.
```
Clarified:

```
// This leads to a short downtime for operations that need the driver:
//   - Pods cannot start unless the claims they depend on were already
//     prepared for use.
//   - Cleanup after the last pod which used a claim gets delayed
//     until the driver is available again. The pod is not marked
//     as terminated. This prevents reusing the resources used by
//     the pod for other pods.
//   - Running pods are *not* affected as far as Kubernetes is
//     concerned. However, a DRA driver might provide required runtime
//     services. Vendors need to document this.
//
// Note that the second point also means that draining a node should
// first evict normal pods, then the driver DaemonSet Pod.
```

This is a reference to the problem discussed in https://github.com/kubernetes/kubernetes/issues/129402.

pohly commented:
I had to rebase to take advantage of https://github.com/kubernetes/kubernetes/pull/130747.

Please see the [intermediate update](https://github.com/kubernetes/kubernetes/compare/57ec0dc0631aec882f60eaee7225eb38ba49c2a5..7469a9d468ed1c09e4387ca65202e905782a183f) for the manual changes.

pohly commented:
I think all review comments have been resolved. As soon as @elezar confirms that, we can merge?

@elezar, do you want to do the final LGTM?

pohly commented:
/hold cancel

byako reviewed (approved):
lgtm

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -263,27 +364,24 @@ type draPlugin struct {
 	logger := klog.FromContext(ctx)
 	o := options{
 		logger:        klog.Background(),
 		grpcVerbosity: 6, // Logs requests and responses, which can be large.
```
Out of scope for this PR: Is this a sane default given the comment?

pohly replied:
Yes, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -263,27 +364,24 @@ type draPlugin struct {
 		logger:        klog.Background(),
 		grpcVerbosity: 6, // Logs requests and responses, which can be large.
-		nodeV1alpha4:  true,
+		serialize:     true,
```
Another convention question. Is it common to sort field initialization to match declarations / alphabetically?

pohly replied:
Definitely not alphabetical. According to declaration makes more sense, although sometimes declarations follow a grouping that might make less for initialization. I would say "whatever looks good" is fine.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -294,22 +392,25 @@ func Start(ctx context.Context, nodeServers []interface{}, opts ...Option) (resu
 	}
-	if o.pluginRegistrationEndpoint == emptyEndpoint {
-		return nil, errors.New("a Unix domain socket path and/or listener must be set for the registrar")
+	draEndpoint := endpoint{
```
nit: Since we're moving from defaults to actual initialization, should this be moved further down? Also, does it make sense to instantiate `draEndpoint` just before we need it? Would this make factoring code out later easier?

pohly replied:
Moved.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -337,39 +438,23 @@ func Start(ctx context.Context, nodeServers []interface{}, opts ...Option) (resu
-
 	// Now make it available to kubelet.
-	registrar, err := startRegistrar(klog.NewContext(ctx, klog.LoggerWithName(logger, "registrar")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, supportedServices, o.draAddress, o.pluginRegistrationEndpoint)
+	registrar, err := startRegistrar(klog.LoggerWithName(logger, "registrar"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, supportedServices, path.Join(draEndpoint.dir, draEndpoint.file), o.pluginRegistrationEndpoint)
```
does adding a `Path` function to the `endpoint` type make sense? Not sure how commonplace the `Join` of `.dir` and `.file` is though.

pohly replied:
It occurred in a few places, so I added such a method.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -456,12 +564,120 @@ func (d *draPlugin) PublishResources(_ context.Context, resources resourceslice.
+	return d.grpcMutex.Unlock
+}
+
+// nodePluginImplementation is a thin wrapper around the helper instance.
```
Since these are intended to be private, does it make things cleaner to also move them to a separate `go` file?

pohly replied:
After dropping the public interface for the helper, the public API is spread across the entire file. An interface allows documenting the full public API in a separate file, but it's not good practice. For example, extending that interface is an API break because there might be other implementations of it, unless extra care is taken with adding an unexported method.

If someone wants to learn how to use this package, they should use `go doc`. For someone implementing it I want to have it in one file.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -456,12 +564,120 @@ func (d *draPlugin) PublishResources(_ context.Context, resources resourceslice.
+		return nil, fmt.Errorf("get resource claims: %w", err)
+	}
+
+	// Serialize if enabled, otherwise NOP.
+	defer d.serializeGRPC()()
```
Instead of a comment each time, should we call the function `serializeGRPCIfEnabled()`?

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/noderegistrar.go:
```diff
@@ -30,17 +30,15 @@ type nodeRegistrar struct {
-//
-// The context is only used for additional values, cancellation is ignored.
-func startRegistrar(valueCtx context.Context, grpcVerbosity int, interceptors []grpc.UnaryServerInterceptor, streamInterceptors []grpc.StreamServerInterceptor, driverName string, supportedServices []string, endpoint string, pluginRegistrationEndpoint endpoint) (*nodeRegistrar, error) {
+func startRegistrar(logger klog.Logger, grpcVerbosity int, interceptors []grpc.UnaryServerInterceptor, streamInterceptors []grpc.StreamServerInterceptor, driverName string, supportedServices []string, pluginEndpoint string, pluginRegistrationEndpoint endpoint) (*nodeRegistrar, error) {
```
Question: why is `pluginEndpoint` not an `endpoint` type? Does this mean that the argument should be renamed or the type changed?

pohly replied:
`registerapi.PluginInfo` as an `Endpoint string` field. This name here reflects that more than the rest of the code where it normally would be called "socket path". I've switched to that.

elezar commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/nonblockinggrpcserver.go:
```diff
@@ -36,60 +37,67 @@ type grpcServer struct {
-			return nil, fmt.Errorf("remove Unix domain socket: %v", err)
+	socketpath := path.Join(endpoint.dir, endpoint.file)
+	listen := endpoint.listen
+	if listen == nil {
```
Should this logic be attached to the `endpoint` type?

pohly replied:
Yes. I tried it and it made the code a bit more readable.

elezar commented on test/e2e/dra/deploy.go:
```diff
@@ -419,6 +426,7 @@ func (d *Driver) SetUp(nodes *Nodes, resources Resources, devicesPerNode ...map[
 		} else {
 			fileOps.NumDevices = numDevices
 		}
+		listenerPort := int32(9000)
```
We previously had two ports. Was this change intentional?

pohly replied:
This is now just the starting value for all ports.

pohly replied:
Let me add a comment...

elezar commented on test/e2e/dra/deploy.go:
```diff
@@ -548,21 +558,161 @@ func (d *Driver) podIO(pod *v1.Pod) proxy.PodDirIO {
+func listen(f *framework.Framework, pod *v1.Pod, port *int32) func(ctx context.Context, path string) (net.Listener, error) {
+	return func(ctx context.Context, path string) (l net.Listener, e error) {
+		// "Allocate" a new port by by bumping the per-pod counter by one.
+		port := atomic.AddInt32(port, 1)
```
Or does this makes sure that the ports are unique?

pohly replied:
Yes :smile:

pohly commented:
@elezar: please take another look at [the changes](https://github.com/kubernetes/kubernetes/compare/6b3584f2cdefd01d814e094515ad554b30c8e347..9f0d197f48fe6f1e1163a70998306f01b5912a18).

elezar reviewed (approved):
Thanks @pohly.

/lgtm
